pep-0677.rst

```diff
@@ -0,0 +1,854 @@
+Many languages use the same arrow token for both lambdas and callable
+types, but Python is unique in that types are expressions and have to
+evaluate to runtime values. Our view is that this merits using
+separate tokens, and given the existing use of ``->`` for return types
+in function signatures it would be more coherent to use ``->`` for
+callable types and ``=>`` for lambdas.
+
+Runtime Behavior
+----------------
+
+Our tentative plan is that:
+
+- The ``__repr__`` will show an arrow syntax literal.
+- We will provide a new API where the runtime data structure can be
+  accessed in the same manner as the AST data structure.
+- We will ensure that we provide an API that is backward-compatible
+  with ``typing.Callable`` and ``typing.Concatenate``, specifically
+  the behavior of ``__args__`` and ``__parameters__``.
+
+Because these details are still under debate we are currently
+maintaining `a separate doc
+<https://docs.google.com/document/d/15nmTDA_39Lo-EULQQwdwYx_Q1IYX4dD5WPnHbFG71Lk/edit>`_
+with details about the new builtins, the evaluation model, how to
+provide both a backward-compatible and more structured API, and
+possible alternatives to the current plan.
+
+Once the plan is finalized we will include a full specification of
+runtime behavior in this section of the PEP.
+
+Rejected Alternatives
+=====================
+
+Many of the alternatives we considered would have been more expressive
+than ``typing.Callable``, for example adding support for describing
+signatures that include named, optional, and variadic arguments.
+
+To determine which features we most needed to support with a callable
+type syntax, we did an extensive analysis of existing projects:
+
+- `stats on the use of the Callable type <https://github.com/pradeep90/annotation_collector#typed-projects---callable-type>`_;
+- `stats on how untyped and partially-typed callbacks are actually used <https://github.com/pradeep90/annotation_collector#typed-projects---callback-usage>`_.
+
+We decided on a simple proposal with improved syntax for the existing
+``Callable`` type because the vast majority of callbacks can be correctly
+described by the existing ``typing.Callable`` semantics:
+
+- Positional parameters: By far the most important case to handle well
+  is simple callable types with positional parameters, such as
+  ``(int, str) -> bool``
+- ParamSpec and Concatenate: The next most important feature is good
+  support for PEP 612 ``ParamSpec`` and ``Concatenate`` types like
+  ``(**P) -> bool`` and ``(int, **P) -> bool``. These are common
+  primarily because of the heavy use of decorator patterns in python
+  code.
+- TypeVarTuples: The next most important feature, assuming PEP 646 is
+  accepted, is for unpacked types which are common because of cases
+  where a wrapper passes along ``*args`` to some other function.
+
+Features that other, more complicated proposals would support account
+for fewer than 2% of the use cases we found. These are already
```

Comment:
I wish we had more data to support the 2% number, since I still have a nagging doubt about it. I worry that this number will be quoted endlessly regardless of its truth value. Maybe Pradeep can publish a set of raw data sources one can use to repeat his analysis? (It would be cool if we could repeat the same analysis at some future point and see if the number is changed.)

Reply:
[Guido]

I'd already shared the [script](https://github.com/pradeep90/annotation_collector#calls-to-callback-parameters) to repeat the analysis on any project from the command-line. The list of OSS projects with stats is [here](https://github.com/pradeep90/annotation_collector#stats) (raw [callables](https://github.com/pradeep90/annotation_collector/tree/master/data)). I definitely want people to repeat the analysis on their favorite code bases.

(@stroxler Could we explicitly link to the script in the resources and maybe drop a couple of lines describing the methods, the assortment of projects, and the limitations? We're referring to the stats quite a bit and people might be curious.)

[Guido]

> It would be cool if we could repeat the same analysis at some future point and see if the number is changed.

Yes, we can rerun the above script later on to see how usage has changed.

> I worry that this number will be quoted endlessly

+1. Feel like we should qualify it instead of presenting it as a bare stat, especially since people will question it. Maybe something like "X% of the cases we found in the N open-source projects we analyzed". That makes it explicit that YMMV and that this is an average.

> since I still have a nagging doubt about it

Were there particular types of projects where you feel the distribution would be different? It can help to test the hypothesis directly instead of letting the doubt nag :)

(For example, I had been curious how untyped projects used callbacks, which is why I included untyped projects like django. Same for open-source vs internal. Steven had been curious whether application code differed much from library code.)

Reply:
I need to give up on fighting this on a numeric basis. I agree that it's probably little-used.

I still worry about confirmation bias, but my biggest, intuitive objection is that I feel the extended form is needed for *completeness* (i.e., there are function types whose values you can spell with a single `def` but whose type must be spelled with a callback protocol). I know it's fashionable to make fun of completionism, but I suspect that a lot of other useful features wouldn't have been in the language if we had based our decisions purely on data about current usage.

My second objection is that callback protocols are both ugly and unintuitive.